GH-132983: Build ``_zstd`` on Windows

AA-Turner opened the pull request:
Per the documentation in [`zlib/lib`](https://github.com/facebook/zstd/tree/dev/lib#readme), we build `lib/common`, `lib/compress`, `lib/decompress`, and `lib/dictBuilder`. 

Building `lib/legacy` is noted as optional to support decompression only of archives created with Zstd 0.8 (2016) or earlier. That release is 10 years old, so for simplicity I chose not to build it, but we could add it in if thought useful.

A

<!-- gh-issue-number: gh-132983 -->
* Issue: gh-132983
<!-- /gh-issue-number -->

AA-Turner committed:
Sorting

AA-Turner committed:
Zstandard scaffold

AA-Turner committed:
Add Zstandard projects

emmatyping reviewed (approved):
Looks good!

emmatyping commented on PCbuild/pcbuild.sln:
```diff
@@ -633,6 +637,38 @@ Global
 		{900342D7-516A-4469-B1AD-59A66E49A25F}.Release|Win32.Build.0 = Release|Win32
 		{900342D7-516A-4469-B1AD-59A66E49A25F}.Release|x64.ActiveCfg = Release|x64
 		{900342D7-516A-4469-B1AD-59A66E49A25F}.Release|x64.Build.0 = Release|x64
+		{4D7C112F-3083-4D9E-9754-9341C14D9B39}.Debug|ARM.ActiveCfg = Debug|ARM
```
Do you know why this was reordered? I don't think it matters much but to minimize the diff it might be nice  to restore the ordering (though maybe the ordering was wrong before!)

AA-Turner replied:
I just ran the commands in VS and it re-ordered every time. I'll look into splitting the ordering change into a new PR.

AA-Turner replied:
Conflicts resolved

zooba commented on PCbuild/pcbuild.sln:
```diff
@@ -165,6 +165,10 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "zlib-ng", "zlib-ng.vcxproj"
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "_zstd", "_zstd.vcxproj", "{07029B86-F3E9-443E-86FB-78AA6D47FED1}"
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "zstd", "zstd.vcxproj", "{8D80E018-6A7D-4B46-A7BE-D9E02A20D473}"
+EndProject
```
There should be some project dependency information in here - SLN builds don't use the `ProjectReference` elements from the projects, unfortunately. Easiest way is to open in VS, right-click the project and choose "Project Dependencies" (on `_zstd` to select `zstd` I'd guess, and then on `python` to select `_zstd`)

AA-Turner replied:
Done: [`b9208b6` (#133366)](https://github.com/python/cpython/pull/133366/commits/b9208b60b429c35482a398a61f973e3dfbd8dc57)

zooba reviewed (commented):
Looks fine to me, though I'm not sure we need a separate project for `zstd`? We should be able to reference all the files directly through `_zstd` (we couldn't for zlib-ng because their build process is way too complex to integrate into `pythoncore`, but this one looks pretty simple _and_ already has its own extension module that's separate).

sethmlarson reviewed (approved):
SBOM changes LGTM, thank you!

AA-Turner committed:
Add dependency of _zstd to python

AA-Turner committed:
Combine zstd.vcxproj into _zstd.vcxproj

zooba commented on PCbuild/python.props:
```diff
@@ -87,6 +87,7 @@
     <nasmDir Condition="$(nasmDir) == ''">$(ExternalsDir)\nasm-2.11.06\</nasmDir>
     <zlibDir Condition="$(zlibDir) == ''">$(ExternalsDir)\zlib-1.3.1\</zlibDir>
     <zlibNgDir Condition="$(zlibNgDir) == ''">$(ExternalsDir)\zlib-ng-2.2.4\</zlibNgDir>
+    <zstdDir Condition="$(zstdDir) == ''">$(ExternalsDir)\zstd-1.5.7\lib\</zstdDir>
```
Sorry to be a pain, just spotted this. Can we move the `lib\` from here to everywhere that references this? (Or another variable that is based on this one).

These variables are quasi-public API, in that they're deliberately overridable by builders to use alternate dependencies. We should keep them all following the same pattern - that is, pointing to the root of the sources.

AA-Turner replied:
Done: [`1ae3a76` (#133366)](https://github.com/python/cpython/pull/133366/commits/1ae3a76e4827f2c8c005f92697683ca87d97fe5e)

AA-Turner committed:
Rename `Modules\zdict.c` to `zstddict.c`

MSB8027: Two or more files with the name of zdict.c will produce outputs to the same location. This can lead to an incorrect build result.  The files involved are ..\Modules\_zstd\zdict.c, \externals\zstd-1.5.7\lib\dictBuilder\zdict.c.

AA-Turner committed:
Point $(zstdDir) to the root

zooba reviewed (approved):
LGTM. Merge whenever you're ready
